modules/bm_sim/src/lookup_structures.cpp

```diff
@@ -0,0 +1,247 @@
+    trie = bf_lpm_trie_create(key_width_bytes, true);
+  }
+
+ private:
+  size_t key_width_bytes{0};
+  bf_lpm_trie_t *trie{nullptr};
+};
+
+class ExactMap : public LookupStructure<ExactMatchKey> {
+ public:
+  explicit ExactMap(size_t size) {
+    entries_map.reserve(size);
+  }
+
+  bool lookup(const ByteContainer & key,
+              internal_handle_t * handle) override {
+    const auto entry_it = entries_map.find(key);
+    if (entry_it == entries_map.end()) {
+      return false;  // Nothing found
+    } else {
+      *handle = entry_it->second;
+      return true;
+    }
+  }
+
+  bool entry_exists(const ExactMatchKey & key) override {
+    (void) key;
+    return entries_map.find(key.data) != entries_map.end();
+  }
+
+  void store_entry(const ExactMatchKey & key,
+      internal_handle_t handle) override {
+    entries_map[key.data] = handle;  // key is copied, which is not great
+  }
+
+  void delete_entry(const ExactMatchKey & key) override {
+    entries_map.erase(key.data);
+  }
+
+  void clear() override {
+    entries_map.clear();
+  }
+
+ private:
+  // TODO(gordon) Should this be internal_handle_t ?
+  std::unordered_map<ByteContainer, entry_handle_t, ByteContainerKeyHash>
+    entries_map{};
+};
+
+class TernaryMap : public LookupStructure<TernaryMatchKey> {
+ public:
+  explicit TernaryMap(size_t nbytes_key)
+    : nbytes_key(nbytes_key) {}
+
+  bool lookup(const ByteContainer & key_data,
+      internal_handle_t * handle) override {
+    int min_priority = std::numeric_limits<int>::max();;
+    bool match;
+
+    const TernaryMatchKey *entry;
+    const TernaryMatchKey *min_entry = nullptr;
+    entry_handle_t min_handle = 0;
+
+    for (auto it = this->handles.begin(); it != this->handles.end(); ++it) {
+      entry = &std::get<0>(*it);
+
+      if (entry->priority >= min_priority) continue;
+
+      match = true;
+      for (size_t byte_index = 0; byte_index < this->nbytes_key; byte_index++) {
+        if (entry->data[byte_index] !=
+            (key_data[byte_index] & entry->mask[byte_index])) {
+          match = false;
+          break;
+        }
+      }
+
+      if (match) {
+        min_priority = entry->priority;
+        min_entry = entry;
+        min_handle = std::get<1>(*it);
+      }
+    }
+
+    if (min_entry) {
+      *handle = min_handle;
+      return true;
+    }
+
+    return false;
+  }
+
+  bool entry_exists(const TernaryMatchKey & key) override {
+    auto it = find_handle(key);
+
+    return it != this->handles.end();
+  }
+
+  void store_entry(const TernaryMatchKey & key,
+      internal_handle_t handle) override {
+    // TODO(gordon) avoid copying key ?
+    handles.emplace_back(std::make_tuple(key, handle));
+  }
+
+  void delete_entry(const TernaryMatchKey & key) override {
+    auto it = find_handle(key);
+    this->handles.erase(it);
+  }
+
+  void clear() override {
+    this->handles.clear();
+  }
+
+ private:
+  std::vector<std::tuple<TernaryMatchKey, internal_handle_t>> handles;
```

Comment:
the internal handle can be used as and index into the vector (this is actually the difference with the regular client-facing handle, which includes a version number); so I would just use that to make the code simpler. Don't forget to initialize your vector to `size` elements.
`handles` is a bad name for this (especially after removing the handles)